### PR TITLE
Sanitize newprefix

### PR DESCRIPTION
**Pull request by majst01**
@kolsa found a important case for misusage.

Before this it was possible to create two prefixes without any complains from this library:

`10.0.0.0/24` and `10.0.0.1/24`
because the prefix cidr is the primary key in the storage, but it was not checked that the second prefix is overlapping the first.
metal-api, which uses this library checked that beforehand, but this should be done here as well.

Through this finding, we identified a second problem which is IPv6 only, related to this:

IPv6 addresses are possible in the full form, and in a condensed form where all zeros are ommited. For example someone could create a prefix `2001:0abc::/64` and later asks for a IP in `2001:abc::/64`, both are the same prefix, but in the current implementation of go-ipam the prefix will not be found because it is stored in the first form.

With this PR the cidr and the parentcidr are parsed and stored in the normalized form with all zeros stripped.

**!!! This is breaking for existing IPv6 use cases !!!**

Additionally the Prefix Cidr is stored in canonical from, this means `10.0.0.1/24` will be stored as `10.0.0.0/24` in the database. This changes the behaviour of existing IPv4 prefixes which where not stored in canonical form.

For collision detection a new storage interface method `ReadAllPrefixCidrs` was introduced to read only the prefix cidrs instead of parsing the all prefixes, which prevents the requirement to read the whole DB for this purpose.

Considerable slower due to the additional prefix parsing and collision detection during NewPrefix:
```
name                        old time/op    new time/op    delta
NewPrefix/Memory-4            1.74µs ±11%    2.83µs ± 9%  +62.67%  (p=0.008 n=5+5)
NewPrefix/Postgres-4          10.7ms ± 4%    11.1ms ± 7%     ~     (p=0.548 n=5+5)
NewPrefix/Cockroach-4         29.6ms ±25%    28.9ms ±26%     ~     (p=0.841 n=5+5)
AcquireIP/Memory-4            2.90µs ± 0%    4.56µs ± 6%  +57.55%  (p=0.008 n=5+5)
AcquireIP/Postgres-4          14.8ms ± 7%    14.6ms ± 7%     ~     (p=0.690 n=5+5)
AcquireIP/Cockroach-4         22.0ms ±11%    20.1ms ± 8%     ~     (p=0.095 n=5+5)
AcquireChildPrefix/8/14-4     6.16µs ± 3%    9.13µs ±10%  +48.11%  (p=0.008 n=5+5)
AcquireChildPrefix/8/16-4     6.12µs ± 0%    8.16µs ± 0%  +33.27%  (p=0.008 n=5+5)
AcquireChildPrefix/8/20-4     6.09µs ± 0%    8.20µs ± 3%  +34.60%  (p=0.016 n=4+5)
AcquireChildPrefix/8/22-4     6.09µs ± 0%    8.23µs ± 3%  +35.15%  (p=0.008 n=5+5)
AcquireChildPrefix/8/24-4     6.07µs ± 0%    8.12µs ± 0%  +33.78%  (p=0.008 n=5+5)
AcquireChildPrefix/16/18-4    6.10µs ± 0%    8.22µs ± 2%  +34.73%  (p=0.008 n=5+5)
AcquireChildPrefix/16/20-4    6.38µs ± 5%    8.47µs ± 4%  +32.81%  (p=0.008 n=5+5)
AcquireChildPrefix/16/22-4    6.08µs ± 0%    8.36µs ± 4%  +37.44%  (p=0.008 n=5+5)
AcquireChildPrefix/16/24-4    6.06µs ± 0%    8.88µs ± 7%  +46.38%  (p=0.008 n=5+5)
AcquireChildPrefix/16/26-4    6.09µs ± 2%    8.59µs ± 4%  +40.88%  (p=0.008 n=5+5)
PrefixOverlapping-4            260ns ± 1%     277ns ± 1%   +6.64%  (p=0.008 n=5+5)

name                        old alloc/op   new alloc/op   delta
NewPrefix/Memory-4            1.54kB ± 0%    1.73kB ± 0%  +12.50%  (p=0.008 n=5+5)
NewPrefix/Postgres-4          5.58kB ± 2%    6.26kB ± 1%  +12.20%  (p=0.008 n=5+5)
NewPrefix/Cockroach-4         5.61kB ± 1%    6.26kB ± 1%  +11.66%  (p=0.008 n=5+5)
AcquireIP/Memory-4            2.36kB ± 0%    2.58kB ± 0%   +9.49%  (p=0.008 n=5+5)
AcquireIP/Postgres-4          10.4kB ± 1%    10.5kB ± 0%   +1.72%  (p=0.008 n=5+5)
AcquireIP/Cockroach-4         10.4kB ± 0%    10.6kB ± 0%   +1.98%  (p=0.008 n=5+5)
AcquireChildPrefix/8/14-4     4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/8/16-4     4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/8/20-4     4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/8/22-4     4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/8/24-4     4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/16/18-4    4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/16/20-4    4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/16/22-4    4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/16/24-4    4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
AcquireChildPrefix/16/26-4    4.64kB ± 0%    5.08kB ± 0%   +9.50%  (p=0.008 n=5+5)
PrefixOverlapping-4            0.00B          0.00B          ~     (all equal)

name                        old allocs/op  new allocs/op  delta
NewPrefix/Memory-4              20.0 ± 0%      27.0 ± 0%  +35.00%  (p=0.008 n=5+5)
NewPrefix/Postgres-4             123 ± 0%       144 ± 0%  +17.07%  (p=0.008 n=5+5)
NewPrefix/Cockroach-4            123 ± 0%       144 ± 0%  +17.07%  (p=0.016 n=4+5)
AcquireIP/Memory-4              42.0 ± 0%      52.0 ± 0%  +23.81%  (p=0.008 n=5+5)
AcquireIP/Postgres-4             249 ± 0%       259 ± 0%   +4.02%  (p=0.016 n=4+5)
AcquireIP/Cockroach-4            250 ± 0%       260 ± 0%   +4.00%  (p=0.029 n=4+4)
AcquireChildPrefix/8/14-4       72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/8/16-4       72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/8/20-4       72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/8/22-4       72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/8/24-4       72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/16/18-4      72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/16/20-4      72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/16/22-4      72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/16/24-4      72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
AcquireChildPrefix/16/26-4      72.0 ± 0%      89.0 ± 0%  +23.61%  (p=0.008 n=5+5)
PrefixOverlapping-4             0.00           0.00          ~     (all equal)
```

### CODE REVIEW

**Comment by majst01** [comment]
We really need to store only in the sanitized form, for both cidr  and parentcidr.

**Review by kolsa — APPROVED** [review]
LGTM

File: prefix.go [hunk 1]
```diff
@@ -140,10 +140,18 @@ type Usage struct {
...
 func (i *ipamer) NewPrefix(cidr string) (*Prefix, error) {
+	existingPrefixes, err := i.storage.ReadAllPrefixCidrs()
+	if err != nil {
+		return nil, err
+	}
 	p, err := i.newPrefix(cidr, "")
 	if err != nil {
 		return nil, err
 	}
+	err = i.PrefixesOverlapping(existingPrefixes, []string{p.Cidr})
```
Comment (droid42) [inline prefix.go:151]: Without lock over all statements (getAllPrefixes, checkOverlap and createPrefix) this is not threadsafe and duplicates can still occur.

**Review by droid42 — CHANGES_REQUESTED** [review]
We should consider to make NewPrefix() threadsafe by introducing a lock, to prevent duplicates - even if this will hurt the performance significantly.

File: prefix.go (hunk 1 above)
Comment (majst01) [reply prefix.go:153]: Done, still thinking if we should, albeit breaks API, remove the ipamer.PrefixFrom Signature and do all work in ipam.NewPrefix.

**Review by droid42 — APPROVED** [review]
LGTM

**Comment by majst01** [comment]
Update benchmark accordingly, nothing shocking.

**Pull request merged**